File: library/src/main/java/name/zeno/android/exception/ZException.java
package name.zeno.android.exception;

import android.os.Build;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.util.Date;

import name.zeno.android.util.ZDate;

/**
 * Create Date: 16/6/3
 *
 * @author 陈治谋 (513500085@qq.com)
 */
@SuppressWarnings("unused")
public class ZException extends RuntimeException
{
  public static final int ERR_DEFAULT        = 1;
  public static final int ERR_SERVICE        = 2;
  public static final int ERR_NETWORK_ON_UI  = 3;
  public static final int ERR_UNKNOWN_HOST   = 4;
  public static final int ERR_DATA_PARSE     = 5;
  public static final int ERR_TIMEOUT        = 6;
  public static final int ERR_CONNECT_FAILED = 7;
  public static final int ERR_NOT_FOUND      = 8;

  private int errCode;

  public ZException(int errCode)
  {
    this.errCode = errCode;
  }

  public ZException(int errCode, String detailMessage)
  {
    super(detailMessage);
    this.errCode = errCode;
  }

  public ZException(int errCode, String detailMessage, Throwable throwable)
  {
    super(detailMessage, throwable);
    this.errCode = errCode;
  }

  public ZException(Throwable throwable)
  {
    super(throwable);
  }

  public static String info(Throwable throwable)
  {
    StringBuilder sb;
    sb = new StringBuilder()
        .append("生产厂商：").append(Build.MANUFACTURER).append("\n")
        .append("手机型号：").append(Build.MODEL).append("\n")
        .append("系统版本：").append(Build.VERSION.RELEASE).append("\n")
        .append("异常时间：").append(ZDate.nowString("yyyy/MM/dd HH:mm:ss")).append("\n\n")
        .append("异常类型：").append(throwable.getClass().getName()).append("\n")
        .append("异常信息：").append(throwable.getMessage()).append("\n\n")
        .append("异常堆栈：\n").append(stack(throwable));
    return sb.toString();
  }

  public static String stack(Throwable throwable)
  {
    //构建字符串
    Writer      writer      = new StringWriter();
    PrintWriter printWriter = new PrintWriter(writer);
    throwable.printStackTrace(printWriter);
    Throwable cause = throwable.getCause();
    while (cause != null) {
      cause.printStackTrace(printWriter);
      cause = cause.getCause();
    }

    return writer.toString();
  }

}

File: zako/src/main/java/cn/izeno/android/exception/ZException.kt
package cn.izeno.android.exception

import android.os.Build
import cn.izeno.android.util.ZDate
import java.io.PrintWriter
import java.io.StringWriter

/**
 * @author 陈治谋 (513500085@qq.com)
 * @since  16/6/3
 */
@Suppress("unused")
class ZException : RuntimeException {

  var errCode: Int = 0

  constructor() : super()
  constructor(message: String?) : super(message)
  constructor(message: String?, cause: Throwable?) : super(message, cause)
  constructor(cause: Throwable?) : super(cause)
  constructor(message: String?, cause: Throwable?, enableSuppression: Boolean, writableStackTrace: Boolean) : super(message, cause, enableSuppression, writableStackTrace)

  fun code(code: Int) = apply {
    errCode = code
  }

  fun info() = info(this)

  companion object {
    const val ERR_DEFAULT = 1
    const val ERR_SERVICE = 2
    const val ERR_NETWORK_ON_UI = 3
    const val ERR_UNKNOWN_HOST = 4
    const val ERR_DATA_PARSE = 5
    const val ERR_TIMEOUT = 6
    const val ERR_CONNECT_FAILED = 7
    const val ERR_NOT_FOUND = 8

    fun info(throwable: Throwable): String = StringBuilder()
        .appendln("生产厂商：${Build.MANUFACTURER}")
        .appendln("手机型号：${Build.MODEL}")
        .appendln("系统版本：${Build.VERSION.RELEASE}")
        .appendln("异常时间：${ZDate.nowString("yyyy/MM/dd HH:mm:ss")}")
        .appendln("异常类型：${throwable.javaClass.name}")
        .appendln("异常信息：${throwable.message}")
        .appendln("异常堆栈：")
        .append(stack(throwable))
        .toString()

    fun stack(throwable: Throwable): String {
      //构建字符串
      val writer = StringWriter()
      val printWriter = PrintWriter(writer)
      throwable.printStackTrace(printWriter)
      var cause: Throwable? = throwable.cause
      while (cause != null) {
        cause.printStackTrace(printWriter)
        cause = cause.cause
      }

      return writer.toString()
    }
  }

}
